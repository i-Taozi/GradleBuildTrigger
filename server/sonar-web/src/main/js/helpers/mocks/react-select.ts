/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { GroupTypeBase, OptionProps } from 'react-select';

export function mockReactSelectOptionProps<
  OptionType,
  IsMulti extends boolean,
  GroupType extends GroupTypeBase<OptionType> = GroupTypeBase<OptionType>
>(
  data: OptionType,
  overrides?: OptionProps<OptionType, IsMulti, GroupType>
): OptionProps<OptionType, IsMulti, GroupType> {
  return {
    ...overrides,
    data
  } as OptionProps<OptionType, IsMulti, GroupType>;
}
