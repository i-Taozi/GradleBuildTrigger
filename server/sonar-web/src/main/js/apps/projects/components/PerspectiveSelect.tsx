/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { omit } from 'lodash';
import * as React from 'react';
import { components, OptionProps } from 'react-select';
import Select from '../../../components/controls/Select';
import BubblesIcon from '../../../components/icons/BubblesIcon';
import ListIcon from '../../../components/icons/ListIcon';
import { translate } from '../../../helpers/l10n';
import { VIEWS, VISUALIZATIONS } from '../utils';

interface Props {
  className?: string;
  onChange: (x: { view: string; visualization?: string }) => void;
  view: string;
  visualization?: string;
}

export interface PerspectiveOption {
  type: string;
  value: string;
  label: string;
}

export default class PerspectiveSelect extends React.PureComponent<Props> {
  handleChange = (option: PerspectiveOption) => {
    if (option && option.type === 'view') {
      this.props.onChange({ view: option.value });
    } else if (option && option.type === 'visualization') {
      this.props.onChange({ view: 'visualizations', visualization: option.value });
    }
  };

  perspectiveOptionRender = (props: OptionProps<PerspectiveOption, false>) => {
    const { data, className } = props;
    return (
      <components.Option
        {...omit(props, ['children', 'className'])}
        className={`it__projects-perspective-option-${data.value} ${className}`}>
        {data.type === 'view' && <ListIcon className="little-spacer-right" />}
        {data.type === 'visualization' && <BubblesIcon className="little-spacer-right" />}
        {props.children}
      </components.Option>
    );
  };

  render() {
    const { view, visualization } = this.props;
    const perspective = view === 'visualizations' ? visualization : view;
    const options: PerspectiveOption[] = [
      ...VIEWS.map(opt => ({
        type: 'view',
        value: opt.value,
        label: translate('projects.view', opt.label)
      })),
      ...VISUALIZATIONS.map(opt => ({
        type: 'visualization',
        value: opt,
        label: translate('projects.visualization', opt)
      }))
    ];
    return (
      <div className={this.props.className}>
        <label id="aria-projects-perspective">{translate('projects.perspective')}:</label>
        <Select
          aria-labelledby="aria-projects-perspective"
          className="little-spacer-left input-medium it__projects-perspective-select"
          isClearable={false}
          onChange={this.handleChange}
          components={{
            Option: this.perspectiveOptionRender
          }}
          options={options}
          isSearchable={false}
          value={options.find(option => option.value === perspective)}
        />
      </div>
    );
  }
}
