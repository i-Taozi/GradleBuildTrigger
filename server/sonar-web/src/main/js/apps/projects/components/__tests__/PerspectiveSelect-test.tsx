/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import BubblesIcon from '../../../../components/icons/BubblesIcon';
import ListIcon from '../../../../components/icons/ListIcon';
import { mockReactSelectOptionProps } from '../../../../helpers/mocks/react-select';
import PerspectiveSelect from '../PerspectiveSelect';

it('should render correctly', () => {
  expect(shallow(<PerspectiveSelect onChange={jest.fn()} view="overall" />)).toMatchSnapshot();
});

it('should render with coverage selected', () => {
  expect(shallowRender({ view: 'visualizations', visualization: 'coverage' })).toMatchSnapshot();
});

it('should render option correctly', () => {
  const wrapper = shallowRender();
  const OptionRender = wrapper.instance().perspectiveOptionRender;
  let option = shallow(
    <OptionRender {...mockReactSelectOptionProps({ type: 'view', value: 'test', label: 'test' })} />
  );

  expect(option.find(ListIcon).type).toBeDefined();
  option = shallow(
    <OptionRender
      {...mockReactSelectOptionProps({ type: 'visualization', value: 'test', label: '' })}
    />
  );
  expect(option.find(BubblesIcon).type).toBeDefined();
});

it('should handle perspective change correctly', () => {
  const onChange = jest.fn();
  const instance = shallowRender({ onChange }).instance();
  instance.handleChange({ label: 'overall', value: 'overall', type: 'view' });
  instance.handleChange({ label: 'leak', value: 'leak', type: 'view' });
  instance.handleChange({ label: 'coverage', value: 'coverage', type: 'visualization' });
  expect(onChange.mock.calls).toMatchSnapshot();
});

function shallowRender(overrides: Partial<PerspectiveSelect['props']> = {}) {
  return shallow<PerspectiveSelect>(
    <PerspectiveSelect onChange={jest.fn()} view="visualizations" {...overrides} />
  );
}
