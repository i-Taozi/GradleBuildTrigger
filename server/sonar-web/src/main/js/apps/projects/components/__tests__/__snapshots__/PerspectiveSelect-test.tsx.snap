// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle perspective change correctly 1`] = `
Array [
  Array [
    Object {
      "view": "overall",
    },
  ],
  Array [
    Object {
      "view": "leak",
    },
  ],
  Array [
    Object {
      "view": "visualizations",
      "visualization": "coverage",
    },
  ],
]
`;

exports[`should render correctly 1`] = `
<div>
  <label
    id="aria-projects-perspective"
  >
    projects.perspective
    :
  </label>
  <Select
    aria-labelledby="aria-projects-perspective"
    className="little-spacer-left input-medium it__projects-perspective-select"
    components={
      Object {
        "Option": [Function],
      }
    }
    isClearable={false}
    isSearchable={false}
    onChange={[Function]}
    options={
      Array [
        Object {
          "label": "projects.view.overall",
          "type": "view",
          "value": "overall",
        },
        Object {
          "label": "projects.view.new_code",
          "type": "view",
          "value": "leak",
        },
        Object {
          "label": "projects.visualization.risk",
          "type": "visualization",
          "value": "risk",
        },
        Object {
          "label": "projects.visualization.reliability",
          "type": "visualization",
          "value": "reliability",
        },
        Object {
          "label": "projects.visualization.security",
          "type": "visualization",
          "value": "security",
        },
        Object {
          "label": "projects.visualization.maintainability",
          "type": "visualization",
          "value": "maintainability",
        },
        Object {
          "label": "projects.visualization.coverage",
          "type": "visualization",
          "value": "coverage",
        },
        Object {
          "label": "projects.visualization.duplications",
          "type": "visualization",
          "value": "duplications",
        },
      ]
    }
    value={
      Object {
        "label": "projects.view.overall",
        "type": "view",
        "value": "overall",
      }
    }
  />
</div>
`;

exports[`should render with coverage selected 1`] = `
<div>
  <label
    id="aria-projects-perspective"
  >
    projects.perspective
    :
  </label>
  <Select
    aria-labelledby="aria-projects-perspective"
    className="little-spacer-left input-medium it__projects-perspective-select"
    components={
      Object {
        "Option": [Function],
      }
    }
    isClearable={false}
    isSearchable={false}
    onChange={[Function]}
    options={
      Array [
        Object {
          "label": "projects.view.overall",
          "type": "view",
          "value": "overall",
        },
        Object {
          "label": "projects.view.new_code",
          "type": "view",
          "value": "leak",
        },
        Object {
          "label": "projects.visualization.risk",
          "type": "visualization",
          "value": "risk",
        },
        Object {
          "label": "projects.visualization.reliability",
          "type": "visualization",
          "value": "reliability",
        },
        Object {
          "label": "projects.visualization.security",
          "type": "visualization",
          "value": "security",
        },
        Object {
          "label": "projects.visualization.maintainability",
          "type": "visualization",
          "value": "maintainability",
        },
        Object {
          "label": "projects.visualization.coverage",
          "type": "visualization",
          "value": "coverage",
        },
        Object {
          "label": "projects.visualization.duplications",
          "type": "visualization",
          "value": "duplications",
        },
      ]
    }
    value={
      Object {
        "label": "projects.visualization.coverage",
        "type": "visualization",
        "value": "coverage",
      }
    }
  />
</div>
`;
